dataflow/src/main/java/org/checkerframework/dataflow/expression/JavaExpressionParseUtil.java

```diff
@@ -0,0 +1,128 @@
+ *     arguments
+ * @checker_framework.manual #dependent-types Annotations whose argument is a Java expression
+ *     (dependent type annotations)
+ */
+public class JavaExpressionParseUtil {
+
+  /** Do not instantiate. */
+  private JavaExpressionParseUtil() {
+    throw new Error("Do not instantiate");
+  }
+
+  /** Regular expression for a formal parameter use. */
+  static final String PARAMETER_REGEX = "#([1-9][0-9]*)";
+
+  /** Pattern for a formal parameter use in "#2" syntax. */
+  static final Pattern PARAMETER_PATTERN = Pattern.compile("#(\\d+)");
+
+  /**
+   * Parsable replacement for formal parameter references. It is parsable because it is a Java
+   * identifier.
+   */
+  static final String PARAMETER_PREFIX = "_param_";
+
+  /** The length of {@link #PARAMETER_PREFIX}. */
+  static final int PARAMETER_PREFIX_LENGTH = PARAMETER_PREFIX.length();
+
+  /** The replacement for a formal parameter in "#2" syntax. */
+  static final String PARAMETER_REPLACEMENT = PARAMETER_PREFIX + "$1";
+
+  /**
+   * Parses a string to a {@link JavaExpression}.
+   *
+   * @param expression the string expression to parse
+   * @param enclosingType type of the class that encloses the JavaExpression
+   * @param thisReference the JavaExpression to which to parse "this", or null if "this" should not
+   *     appear in the expression; not relevant to qualified "SomeClass.this" or
+   *     "package.SomeClass.this"
+   * @param parameters list of JavaExpressions to which to parse formal parameter references such as
+   *     "#2", or null if formal parameter references should not appear in the expression
+   * @param localVarPath if non-null, the expression is parsed as if it were written at this
+   *     location; affects only parsing of local variables
+   * @param pathToCompilationUnit required to use the underlying Javac API
+   * @param env the processing environment
+   * @return {@code expression} as a {@code JavaExpression}
+   * @throws JavaExpressionParseException if the string cannot be parsed
+   */
+  public static JavaExpression parse(
+      String expression,
+      TypeMirror enclosingType,
+      @Nullable ThisReference thisReference,
+      @Nullable List<FormalParameter> parameters,
+      @Nullable TreePath localVarPath,
+      TreePath pathToCompilationUnit,
+      ProcessingEnvironment env)
+      throws JavaExpressionParseException {
+
+    String expressionWithParameterNames =
+        StringsPlume.replaceAll(expression, PARAMETER_PATTERN, PARAMETER_REPLACEMENT);
+    ExpressionTree exprTree;
+    try {
+      JavacParseResult<ExpressionTree> jpr =
+          JavacParse.parseExpression(expressionWithParameterNames);
+      if (jpr.hasParseError()) {
+        throw JavaExpressionParseException.construct(expression, jpr.getParseErrorMessages());
+      }
+      exprTree = jpr.getTree();
+    } catch (IllegalArgumentException e) {
+      @SuppressWarnings("nullness:assignment") // possible real bug?
```

Comment:
`parseExpression` seems to always has a non-null message for the `IllegalArgumentException`s, so this is safe.  But that could change in the future.